File: tests/validate_template.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'pathname'
require 'rexml/document'

REQUIRED_XML = %w[
  content.xml
  styles.xml
  meta.xml
  settings.xml
  META-INF/manifest.xml
].freeze

REQUIRED_PACKAGE_ENTRIES = %w[
  mimetype
  content.xml
  styles.xml
  meta.xml
  settings.xml
  META-INF/manifest.xml
].freeze

EXPECTED_MIMETYPE = "application/vnd.oasis.opendocument.presentation-template"

def fail_with(message)
  warn "ERROR: #{message}"
  exit 1
end

def info(message)
  puts "OK: #{message}"
end

def parse_xml(path)
  File.open(path, "r:utf-8") do |file|
    REXML::Document.new(file)
  end
end

def validate_directory(dir)
  fail_with("No existe el directorio #{dir}") unless dir.directory?

  REQUIRED_XML.each do |relative|
    path = dir.join(relative)
    fail_with("Falta #{relative}") unless path.file?
    parse_xml(path)
    info("XML valido: #{relative}")
  end

  mimetype_path = dir.join("mimetype")
  fail_with("Falta mimetype") unless mimetype_path.file?

  mimetype = mimetype_path.read.strip
  fail_with("mimetype inesperado: #{mimetype}") unless mimetype == EXPECTED_MIMETYPE
  info("mimetype correcto")

  markdown_files = dir.glob("**/*.md").map { |path| path.relative_path_from(dir).to_s }.sort
  if markdown_files.empty?
    info("No hay archivos .md dentro de la plantilla")
  else
    fail_with("Hay archivos .md dentro de la plantilla: #{markdown_files.join(', ')}")
  end
end

def run_command(*command)
  output = `#{command.map { |part| "'#{part.to_s.gsub("'", %q('\\''))}'" }.join(' ')}`
  fail_with("Fallo al ejecutar: #{command.join(' ')}") unless $?.success?
  output
end

def validate_package(otp)
  fail_with("No existe el archivo #{otp}") unless otp.file?

  entries = run_command("unzip", "-Z1", otp.to_s).lines.map(&:strip)
  fail_with("El paquete esta vacio") if entries.empty?
  fail_with("La primera entrada no es mimetype") unless entries.first == "mimetype"

  REQUIRED_PACKAGE_ENTRIES.each do |entry|
    fail_with("Falta #{entry} dentro del .otp") unless entries.include?(entry)
  end
  info("Entradas ODF requeridas presentes en el .otp")

  listing = run_command("unzip", "-lv", otp.to_s)
  mimetype_line = listing.lines.find { |line| line.include?(" mimetype") }
  fail_with("No se pudo inspeccionar mimetype dentro del .otp") unless mimetype_line
  fail_with("mimetype fue comprimido dentro del .otp") unless mimetype_line.include?("Stored")
  info("mimetype va primero y sin compresion en el .otp")
end

input = ARGV[0] or fail_with("Uso: ruby tests/validate_template.rb RUTA_A_PLANTILLA")

dir = Pathname.new(input).expand_path
validate_directory(dir)

otp = dir.parent.join("#{dir.basename}.otp")
if otp.file?
  validate_package(otp)
else
  warn "WARN: No existe #{otp}, se valido solo la carpeta fuente"
end

puts "Plantilla validada: #{dir}"

File: tests/validate_template.py
#!/usr/bin/env python3

from __future__ import annotations

import argparse
from pathlib import Path
import sys
import xml.etree.ElementTree as ET
import zipfile


REQUIRED_XML = [
    "content.xml",
    "styles.xml",
    "meta.xml",
    "settings.xml",
    "META-INF/manifest.xml",
]

REQUIRED_PACKAGE_ENTRIES = [
    "mimetype",
    "content.xml",
    "styles.xml",
    "meta.xml",
    "settings.xml",
    "META-INF/manifest.xml",
]

EXPECTED_MIMETYPE = "application/vnd.oasis.opendocument.presentation-template"


def fail(message: str) -> int:
    print(f"ERROR: {message}", file=sys.stderr)
    return 1


def info(message: str) -> None:
    print(f"OK: {message}")


def validate_directory(template_dir: Path) -> int:
    if not template_dir.is_dir():
        return fail(f"No existe el directorio {template_dir}")

    for relative in REQUIRED_XML:
        path = template_dir / relative
        if not path.is_file():
            return fail(f"Falta {relative}")
        ET.parse(path)
        info(f"XML valido: {relative}")

    mimetype_path = template_dir / "mimetype"
    if not mimetype_path.is_file():
        return fail("Falta mimetype")

    mimetype = mimetype_path.read_text(encoding="utf-8").strip()
    if mimetype != EXPECTED_MIMETYPE:
        return fail(f"mimetype inesperado: {mimetype}")
    info("mimetype correcto")

    markdown_files = sorted(
        str(path.relative_to(template_dir)) for path in template_dir.rglob("*.md")
    )
    if markdown_files:
        return fail(f"Hay archivos .md dentro de la plantilla: {', '.join(markdown_files)}")
    info("No hay archivos .md dentro de la plantilla")
    return 0


def validate_package(otp_path: Path) -> int:
    if not otp_path.is_file():
        return fail(f"No existe el archivo {otp_path}")

    with zipfile.ZipFile(otp_path, "r") as archive:
        infos = archive.infolist()
        if not infos:
            return fail("El paquete esta vacio")
        if infos[0].filename != "mimetype":
            return fail("La primera entrada no es mimetype")

        names = {info.filename for info in infos}
        for entry in REQUIRED_PACKAGE_ENTRIES:
            if entry not in names:
                return fail(f"Falta {entry} dentro del .otp")
        info("Entradas ODF requeridas presentes en el .otp")

        mimetype_info = infos[0]
        if mimetype_info.compress_type != zipfile.ZIP_STORED:
            return fail("mimetype fue comprimido dentro del .otp")
        info("mimetype va primero y sin compresion en el .otp")

    return 0


def main() -> int:
    parser = argparse.ArgumentParser(
        description="Valida una plantilla Impress descomprimida y, si existe, su archivo .otp"
    )
    parser.add_argument("template_dir", help="Ruta al directorio de la plantilla")
    args = parser.parse_args()

    template_dir = Path(args.template_dir).resolve()

    result = validate_directory(template_dir)
    if result != 0:
        return result

    otp_path = template_dir.parent / f"{template_dir.name}.otp"
    if otp_path.exists():
        result = validate_package(otp_path)
        if result != 0:
            return result
    else:
        print(f"WARN: No existe {otp_path}, se valido solo la carpeta fuente", file=sys.stderr)

    print(f"Plantilla validada: {template_dir}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
